packages/playground/data-liberation/src/import/WP_Topological_Sorter.php

```diff
@@ -0,0 +1,461 @@
+		);
+	}
+
+	/**
+	 * Map a category to the index.
+	 *
+	 * @param int $byte_offset The byte offset of the category.
+	 * @param array $data The category data.
+	 */
+	public function map_category( $byte_offset, $data ) {
+		global $wpdb;
+
+		if ( empty( $data ) ) {
+			return false;
+		}
+
+		$category_parent = null;
+
+		if ( array_key_exists( 'parent', $data ) && '' !== $data['parent'] ) {
+			$category_parent = $data['parent'];
+		}
+
+		$wpdb->insert(
+			self::get_table_name(),
+			array(
+				'session_id'   => $this->current_session,
+				'element_type' => self::ELEMENT_TYPE_CATEGORY,
+				'element_id'   => (string) $data['term_id'],
+				'parent_id'    => $category_parent,
+				'byte_offset'  => $byte_offset,
+				// Items with a parent has at least a sort order of 2.
+				'sort_order'   => $category_parent ? 2 : 1,
+			)
+		);
+
+		++$this->total_categories;
+	}
+
+	/**
+	 * Map a post to the index.
+	 *
+	 * @param int $byte_offset The byte offset of the post.
+	 * @param array $data The post data.
+	 */
+	public function map_post( $byte_offset, $data ) {
+		global $wpdb;
+
+		if ( empty( $data ) ) {
+			return false;
+		}
+
+		// No parent, no need to sort.
+		if ( ! isset( $data['post_type'] ) ) {
+			return false;
+		}
+
+		if ( 'post' === $data['post_type'] || 'page' === $data['post_type'] ) {
+			if ( ! $data['post_id'] ) {
+				$this->last_post_id = $this->orphan_post_counter;
+				--$this->orphan_post_counter;
+			}
+
+			$post_parent = null;
+
+			if ( array_key_exists( 'post_parent', $data ) && '0' !== $data['post_parent'] ) {
+				$post_parent = $data['post_parent'];
+			}
+
+			$wpdb->insert(
+				self::get_table_name(),
+				array(
+					'session_id'   => $this->current_session,
+					'element_type' => self::ELEMENT_TYPE_POST,
+					'element_id'   => (string) $data['post_id'],
+					'parent_id'    => $post_parent,
+					'byte_offset'  => $byte_offset,
+					'sort_order'   => $post_parent ? 2 : 1,
+				)
+			);
+
+			++$this->total_posts;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Get the byte offset of an element, and remove it from the list.
+	 *
+	 * @param int $id The ID of the post to get the byte offset.
+	 *
+	 * @return int|bool The byte offset of the post, or false if the post is not found.
+	 */
+	public function get_post_byte_offset( $session_id, $id ) {
+		global $wpdb;
+
+		if ( ! $this->sorted ) {
+			return false;
+		}
+
+		return $wpdb->get_var(
+			$wpdb->prepare(
+				'SELECT byte_offset FROM %i WHERE element_id = %s AND element_type = %d AND session_id = %d LIMIT 1',
+				self::get_table_name(),
+				(string) $id,
+				self::ELEMENT_TYPE_POST,
+				(string) $session_id
+			)
+		);
+	}
+
+	/**
+	 * Get the byte offset of an element, and remove it from the list.
+	 *
+	 * @param string $slug The slug of the category to get the byte offset.
+	 *
+	 * @return int|bool The byte offset of the category, or false if the category is not found.
+	 */
+	public function get_category_byte_offset( $session_id, $slug ) {
```

Comment:
How many methods can we make `private`? Every public method makes a public API and gets covered by the WordPress indefinite BC policy upon merge to WordPress core.